File: product-service/src/main/java/com/intexsoft/webshop/productservice/mapper/CategoryMapper.java
package com.intexsoft.webshop.productservice.mapper;

import com.intexsoft.webshop.productservice.dto.category.CategoryCreateDto;
import com.intexsoft.webshop.productservice.dto.category.CategoryDto;
import com.intexsoft.webshop.productservice.dto.category.CategoryUpdateDto;
import com.intexsoft.webshop.productservice.model.Category;
import org.mapstruct.*;

import java.util.List;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING,
        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
        unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface CategoryMapper {
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "name", source = "name")
    @Mapping(target = "subcategories", ignore = true)
    Category toCategory(CategoryCreateDto categoryCreateDto);

    @Mapping(target = "id", source = "id")
    @Mapping(target = "name", source = "name")
    CategoryDto toCategoryDto(Category category);

    List<CategoryDto> toCategoryDtos(List<Category> categories);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "name", source = "categoryUpdateDto.name")
    @Mapping(target = "subcategories", ignore = true)
    Category updateCategory(@MappingTarget Category category, CategoryUpdateDto categoryUpdateDto);
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/mapper/CategoryMapper.kt
package com.intexsoft.webshop.productservicekt.mapper

import com.intexsoft.webshop.productservicekt.dto.category.CategoryCreateDto
import com.intexsoft.webshop.productservicekt.dto.category.CategoryDto
import com.intexsoft.webshop.productservicekt.dto.category.CategoryUpdateDto
import com.intexsoft.webshop.productservicekt.model.Category
import org.mapstruct.*

@Mapper(
    componentModel = MappingConstants.ComponentModel.SPRING,
    injectionStrategy = InjectionStrategy.CONSTRUCTOR
)
interface CategoryMapper {
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "name", source = "name")
    @Mapping(target = "subcategories", ignore = true)
    fun toCategory(categoryCreateDto: CategoryCreateDto): Category

    @Mapping(target = "id", source = "id")
    @Mapping(target = "name", source = "name")
    fun toCategoryDto(category: Category): CategoryDto
    fun toCategoryDtos(categories: List<Category>): List<CategoryDto>

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "name", source = "categoryUpdateDto.name")
    @Mapping(target = "subcategories", ignore = true)
    fun updateCategory(@MappingTarget category: Category, categoryUpdateDto: CategoryUpdateDto): Category
}